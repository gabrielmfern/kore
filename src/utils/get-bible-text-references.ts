import {
  buildRegExp,
  capture,
  digit,
  optional,
  oneOrMore,
  negated,
  zeroOrMore,
  choiceOf,
  charClass,
  nonWhitespace,
  anyOf,
} from "ts-regex-builder";
import { books, type BookName } from "./books";

const bookName = choiceOf(
  ...books.map((book) => book.name),
  ...books.flatMap((book) => book.alternatives),
);

const whitespace = negated(charClass(nonWhitespace, anyOf("\r\n")));

const referenceSeparator = choiceOf(
  [zeroOrMore(whitespace), ":", zeroOrMore(whitespace)],
  [oneOrMore(whitespace)],
);

const bibleTextRegex = buildRegExp(
  [
    capture(bookName, {
      name: "book",
    }),
    referenceSeparator,
    capture(oneOrMore(digit), {
      name: "chapter",
    }),
    optional([
      referenceSeparator,
      capture(oneOrMore(digit), { name: "verse_start" }),
      optional([
        zeroOrMore(whitespace),
        "-",
        zeroOrMore(whitespace),
        capture(oneOrMore(digit), { name: "verse_end" }),
      ]),
    ]),
  ],
  {
    global: true,
  },
);

export interface TextReference {
  raw: string;
  index: number;

  book: BookName;
  chapter: number;
  verseRange?: [start: number, end: number];
}

export function getBibleTextReferences(content: string) {
  const references: TextReference[] = [];

  for (const match of content.matchAll(bibleTextRegex)) {
    const groups = match.groups;
    if (groups) {
      const { book, chapter, verse_start, verse_end } = groups;
      const properBook = books.find(
        ({ name, alternatives }) =>
          (alternatives as string[]).includes(book) || name === book,
      );
      if (properBook) {
        references.push({
          book: properBook.name,
          index: match.index,
          raw: match[0],
          chapter: Number.parseInt(chapter),
          verseRange: verse_start
            ? [
                Number.parseInt(verse_start),
                verse_end
                  ? Number.parseInt(verse_end)
                  : Number.parseInt(verse_start),
              ]
            : undefined,
        });
      }
    }
  }

  // resets the regex to avoid issues
  bibleTextRegex.lastIndex = 0;

  return references;
}
