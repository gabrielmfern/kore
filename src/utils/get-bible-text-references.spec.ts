import { getBibleTextReferences } from "./get-bible-text-references";

describe("getBibleTextReferences()", () => {
  it("should work with common text references", () => {
    expect(
      getBibleTextReferences("João 3:13 Mateus 8 27 Mateus 14:25 Mateus 16:16"),
    ).toEqual([
      {
        book: "João",
        chapter: 3,
        index: 0,
        raw: "João 3:13",
        verseRange: [13, 13],
      },
      {
        book: "Mateus",
        chapter: 8,
        index: 10,
        raw: "Mateus 8 27",
        verseRange: [27, 27],
      },
      {
        book: "Mateus",
        chapter: 14,
        index: 22,
        raw: "Mateus 14:25",
        verseRange: [25, 25],
      },
      {
        book: "Mateus",
        chapter: 16,
        index: 35,
        raw: "Mateus 16:16",
        verseRange: [16, 16],
      },
    ]);
  });

  it("should work with abbreviations", () => {
    expect(getBibleTextReferences("1Cr 1; 1 Cr 1; 1 Cr 2:3-4")).toEqual([
      {
        book: "1 Crônicas",
        chapter: 1,
        index: 0,
        raw: "1Cr 1",
        verseRange: undefined,
      },
      {
        book: "1 Crônicas",
        chapter: 1,
        index: 7,
        raw: "1 Cr 1",
        verseRange: undefined,
      },
      {
        book: "1 Crônicas",
        chapter: 2,
        index: 15,
        raw: "1 Cr 2:3-4",
        verseRange: [3, 4],
      },
    ]);
  });
});
