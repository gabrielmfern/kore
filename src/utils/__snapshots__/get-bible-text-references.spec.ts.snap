// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getBibleTextReferences() > should work with abbreviations 1`] = `
[
  {
    "book": "1 Crônicas",
    "chapter": 2,
    "index": 0,
    "raw": "1 Cr 2:3-4",
    "verseRange": [
      3,
      4,
    ],
  },
]
`;

exports[`getBibleTextReferences() > should work with common text references 1`] = `
[
  {
    "book": "João",
    "chapter": 3,
    "index": 0,
    "raw": "João 3:13",
    "verseRange": [
      13,
      13,
    ],
  },
  {
    "book": "Mateus",
    "chapter": 8,
    "index": 10,
    "raw": "Mateus 8 27",
    "verseRange": [
      27,
      27,
    ],
  },
  {
    "book": "Mateus",
    "chapter": 14,
    "index": 22,
    "raw": "Mateus 14:25",
    "verseRange": [
      25,
      25,
    ],
  },
  {
    "book": "Mateus",
    "chapter": 16,
    "index": 35,
    "raw": "Mateus 16:16",
    "verseRange": [
      16,
      16,
    ],
  },
]
`;

exports[`getBibleTextReferences() > should work with lower case book names 1`] = `
[
  {
    "book": "Atos",
    "chapter": 4,
    "index": 32,
    "raw": "acts 4:13",
    "verseRange": [
      13,
      13,
    ],
  },
  {
    "book": "Atos",
    "chapter": 4,
    "index": 43,
    "raw": "at 4:10",
    "verseRange": [
      10,
      10,
    ],
  },
  {
    "book": "Atos",
    "chapter": 10,
    "index": 52,
    "raw": "atos 10:14",
    "verseRange": [
      14,
      14,
    ],
  },
]
`;

exports[`getBibleTextReferences() > should work with prefixes on verse 1`] = `
[
  {
    "book": "Atos",
    "chapter": 4,
    "index": 0,
    "raw": "Atos 4 v5",
    "verseRange": [
      5,
      5,
    ],
  },
  {
    "book": "Atos",
    "chapter": 9,
    "index": 11,
    "raw": "Atos 9, verso 4",
    "verseRange": [
      4,
      4,
    ],
  },
  {
    "book": "Atos",
    "chapter": 10,
    "index": 28,
    "raw": "Atos 10   , verso 10",
    "verseRange": [
      10,
      10,
    ],
  },
  {
    "book": "Atos",
    "chapter": 3,
    "index": 50,
    "raw": "Acts 3, verse 3",
    "verseRange": [
      3,
      3,
    ],
  },
  {
    "book": "Atos",
    "chapter": 4,
    "index": 67,
    "raw": "Acts 4 ver 2",
    "verseRange": [
      2,
      2,
    ],
  },
]
`;
